backend/common/src/main/java/ai/verta/modeldb/common/futures/Future.java

```diff
@@ -0,0 +1,387 @@
+    return from(CompletableFuture.failedFuture(ex));
+  }
+
+  public static <U> Future<U> retriableStage(
+      Supplier<Future<U>> supplier, Function<Throwable, Boolean> retryChecker) {
+    final var promise = new CompletableFuture<U>();
+
+    supplier
+        .get()
+        .whenComplete(
+            (value, throwable) -> {
+              boolean retryCheckerFlag;
+              try {
+                retryCheckerFlag = retryChecker.apply(throwable);
+              } catch (Throwable e) {
+                promise.completeExceptionally(
+                    new ExecutionException("retryChecker threw an exception. Not retrying.", e));
+                return;
+              }
+              if (throwable == null) {
+                promise.complete(value);
+              } else if (retryCheckerFlag) {
+                // If we should retry, then perform the retry and map the result of the future to
+                // the current promise
+                // This build up a chain of promises, which can consume memory. I couldn't figure
+                // out how to do better
+                // with Java constraints of final vars in lambdas and not using uninitialized
+                // variables.
+                retriableStage(supplier, retryChecker)
+                    .whenComplete(
+                        (v, t) -> {
+                          if (t == null) {
+                            promise.complete(v);
+                          } else {
+                            promise.completeExceptionally(t);
+                          }
+                        });
+              } else {
+                promise.completeExceptionally(throwable);
+              }
+            });
+
+    return Future.from(promise);
+  }
+
+  public static <U> Future<Optional<U>> flipOptional(Optional<Future<U>> val) {
+    return val.map(future -> future.thenCompose(u -> Future.of(Optional.ofNullable(u))))
+        .orElse(Future.of(Optional.empty()));
+  }
+
+  public T get() throws Exception {
+    try {
+      return stage.toCompletableFuture().get();
+    } catch (ExecutionException ex) {
+      Throwable cause = ex.getCause();
+      if (cause instanceof Exception) {
+        throw (Exception) cause;
+      }
+      throw new ModelDBException(ex);
+    } catch (InterruptedException ex) {
+      // Restore interrupted state...
+      Thread.currentThread().interrupt();
+      throw new ModelDBException(ex);
+    }
+  }
+
+  public CompletionStage<T> toCompletionStage() {
+    return stage;
+  }
+
+  public static void setFutureExecutor(FutureExecutor executor) {
+    if (futureExecutor == null) {
+      futureExecutor = executor;
+    } else {
+      log.warn(
+          "A FutureExecutor has already been set. Ignoring this call.",
+          new RuntimeException("call-site capturer"));
+    }
+  }
+
+  /** @deprecated Only use this as a part of the conversion process between versions of Futures. */
+  @Deprecated
```

Comment:
What does this annotation do? will it make compilation fail?

Reply:
It will generate warnings at compile time by default. And, in IDEA, usages will show up as a warning and ~crossedOut~ in the editor.